#Контрольная работа

##1. Загрузка датасета с помощью библиотеки pandas

In [62]:
# Импорт библиотеки pandas
import pandas as pd

# Загрузка файла
df = pd.read_csv('movies.csv', sep=';')

# Оповестим об успешной загрузке
print("Датасет успешно загружен!")

Датасет успешно загружен!


В начале работы выполнен импорт библиотеки 'pandas'.

Затем выполнена загрузка данных из файла 'movies.csv' при помощи метода 'read_csv()'

##2. Вывод первых 20 строк с помощью метода head()

In [63]:
# Вывод первых 20 строк датасета
df.head(20)

,title,release_year,score,number_of_votes,duration,main_genre,main_production
0,David Attenborough: A Life on Our Planet,2020.0,9.0,31180.0,83,documentary,GB
1,Inception,2010.0,8.8,2268288.0,148,scifi,GB
2,Forrest Gump,1994.0,8.8,1994599.0,142,drama,US
3,Anbe Sivam,2003.0,8.7,20595.0,160,comedy,IN
4,Bo Burnham: Inside,2021.0,8.7,44074.0,87,comedy,US
5,Saving Private Ryan,1998.0,8.6,1346020.0,169,drama,US
6,Django Unchained,2012.0,8.4,1472668.0,165,western,US
7,Dangal,2016.0,8.4,180247.0,161,action,IN
8,Bo Burnham: Make Happy,2016.0,8.4,14356.0,60,comedy,US
9,Louis C.K.: Hilarious,2010.0,8.4,11973.0,84,comedy,US


При помощи метода 'head(20)' прочитаны первые 20 строк таблицы.

##3. Обзор данных - описание столбцов и предметной области


Датасет представляет собой статистическую таблицу с информацией о фильмах их рейтингом и оценками от пользователей.

В таблице указаны основные данные о фильмах: название фильма, жанр фильма, год выхода фильма, страна производства и длительность фильма.

Также в таблице отражены следующие показатели: средняя оценка фильма (его популярность) и количество голосов пользователей, отданных фильму.

Таблица содержит следующие столбцы:

1. **title** - название фильма (основная информация)
2. **release_year** - год выхода фильма (основная информация)
3. **score** - оценка фильма (показатель)
4. **number_of_votes** - количество голосов пользователей (показатель)
5. **duration** - длительность фильма в минутах (основная информация)
6. **main_genre** - основной жанр фильма (основная информация)
7. **main_production** - страна производства (основная информация)

##4. Оценка данных с помощью метода info()

In [64]:
# Получение общей информации о датасете
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 389 entries, 0 to 388
Data columns (total 7 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   title            384 non-null    object 
 1   release_year     387 non-null    float64
 2   score            386 non-null    float64
 3   number_of_votes  388 non-null    float64
 4   duration         389 non-null    int64  
 5   main_genre       389 non-null    object 
 6   main_production  388 non-null    object 
dtypes: float64(3), int64(1), object(3)
memory usage: 21.4+ KB


По информации из метода 'info()' видно, что в таблице всего 7 колонок и 389 строк.

Также видно, что не все столбцы содержат полное количество значений:

- **title** - 5 пропусков (384 из 389)
- **release_year** - 2 пропуска (387 из 389)
- **score** - 3 пропуска (386 из 389)
- **number_of_votes** - 1 пропуск (388 из 389)
- **main_production** - 1 пропуск (388 из 389)

Столбцы '**duration**' и '**main_genre**' не содержат пропусков.

Столбцы таблицы имеют следующие наименования и типы:

1. **title** - название фильма (строка)
2. **release_year** - год выхода фильма (вещественное число)
3. **score** - оценка фильма (вещественное число)
4. **number_of_votes** - количество голосов пользователей (вещественное число)
5. **duration** - длительность фильма в минутах (целое число)
6. **main_genre** - основной жанр фильма (строка)
7. **main_production** - страна производства (строка)

##5. Применение describe() для оценки числовых столбцов

In [65]:
# Статистическое описание числовых столбцов
df.describe()

,release_year,score,number_of_votes,duration
count,387.000000,386.000000,3.880000e+02,389.000000
mean,2010.976744,7.512176,1.391523e+05,123.352185
std,10.805726,0.443870,2.365279e+05,28.304800
min,1954.000000,6.900000,1.013900e+04,28.000000
25%,2007.500000,7.100000,2.055375e+04,104.000000
50%,2014.000000,7.400000,4.606950e+04,122.000000
75%,2018.000000,7.800000,1.551585e+05,139.000000
max,2022.000000,9.000000,2.268288e+06,229.000000


Метод 'df.describe()' выявил 4 столбца с числовыми значениями.

Для удобства описания статистика по каждому столбцу представлена в виде отдельных таблиц.

####1. release_year (год выхода фильма)

| Статистика | Значение |
|------------|----------|
| **count** | 382.0 |
| **mean** | 2011.0 |
| **std** | 10.85 |
| **min** | 1954.0 |
| **25%** | 2008.0 |
| **50%** | 2014.0 |
| **75%** | 2018.0 |
| **max** | 2022.0 |

Из таблицы видно, что:

- Всего числовых значений - 382 (имеется 5 нечисловых значений + 2 пропуска)
- Средний год выпуска - 2011 год
- Отклонение дат выпуска - 11 лет
- Самый старый фильм - 1954 г.
- 25% фильмов выпущены до 2008 года
- Медиана - 2014 год
- 75% фильмов выпущены до 2018 года
- Самый новый фильм - 2022 года

####2. score (оценка фильма)

| Статистика | Значение |
|------------|----------|
| **count** | 382.0 |
| **mean** | 7.51 |
| **std** | 0.44 |
| **min** | 6.90 |
| **25%** | 7.10 |
| **50%** | 7.40 |
| **75%** | 7.80 |
| **max** | 9.00 |

Из таблицы видно, что:

- Всего числовых значений - 382 (имеется 4 нечисловых значения + 3 пропуска)
- Средняя оценка - 7.51
- Отклонение оценок маленькое - 0.44
- Минимальная оценка - 6.9
- 25% фильмов имеют оценку - 7.1
- Медиана - 7.4
- 75% фильмов имеют оценку - 7.8
- Максимальная оценка - 9.0

####3. number_of_votes (количество голосов)

| Статистика | Значение |
|------------|----------|
| **count** | 382.0 |
| **mean** | 137.235 |
| **std** | 235.559 |
| **min** | 10.139 |
| **25%** | 20.471 |
| **50%** | 45.058 |
| **75%** | 153.497 |
| **max** | 2.268.288 |

Из таблицы видно, что:

- Всего числовых значений - 382 (имеется 6 нечисловых значений + 1 пропуск)
- Среднее количество голосов - 137.235
- Отклонение больше среднего (есть выбросы) - 235.559
- Минимум голосов - 10.139
- 25% фильмов имеют голосов - 20.471
- Медиана - 45.058
- 75% фильмов имеют голосов - 153.497
- Максимум голосов - 2.27 млн.

####4. duration (длительность фильма)

| Статистика | Значение |
|------------|----------|
| **count** | 382.0 |
| **mean** | 123.5 |
| **std** | 28.49 |
| **min** | 28.0 |
| **25%** | 104.0 |
| **50%** | 122.0 |
| **75%** | 139.0 |
| **max** | 229.0 |

Из таблицы видно, что:

- Всего числовых значений - 382 (имеется 7 нечисловых значений)
- Средняя длительность - 123.5 мин.
- Отклонение - 28.49 мин.
- Минимум - 28 мин.
- 25% фильмов длятся - 104 мин.
- Медиана - 122 мин.
- 75% фильмов длятся - 139 мин.
- Максимум - 229 мин.

####Итого:

Данные более-менее реалестичны, все числовые столбцы не имеют отрицательных значений, но в каждом столбце имеются нечисловые значения, поэтому значения в строках необходимо привести к одинаковым типам (числовому), либо удалить строки с некорректными данными.
Строки, в которых есть пустые поля с датой выпуска или длительностью фильма лучше удалить, во все остальные пустые (числовые) поля можно подставить 0 после преобразования типов.

#6. Проверка названий столбцов и замена символов

In [66]:
# Вывод названий столбцов
print("Текущие названия столбцов:")
print(df.columns)

# Замена символов на всякий пожарный
df.columns = df.columns.str.replace('с', 'c')
df.columns = df.columns.str.replace('о', 'o')
df.columns = df.columns.str.replace('е', 'e')
df.columns = df.columns.str.replace('а', 'a')
df.columns = df.columns.str.replace('у', 'y')

# Вывод названий столбцов после изменений
print("\nНовые названия столбцов:")
print(df.columns)

Текущие названия столбцов:
Index(['title', 'release_year', 'score', 'number_of_votes', 'duration',
       'main_genre', 'main_production'],
      dtype='object')

Новые названия столбцов:
Index(['title', 'release_year', 'score', 'number_of_votes', 'duration',
       'main_genre', 'main_production'],
      dtype='object')


Из результата работы метода 'columns' видно, что названия не содержат пробелов и приведены все к нижнему регистру.

Для подстраховки все русские буквы 'а', 'е', 'c', 'у' и 'o' были заменены на английские.

#7. Проверка и устранение пропущенных значений, приведение типов

In [67]:
# Подсчёт пропущенных значений по каждому столбцу
print("Количество пропусков в каждом столбце:")
print(df.isna().sum())

# 1. Строки с пустым названием фильма, годом выпуска
# или длительностью фильма удаляем
df = df.dropna(subset=['title', 'release_year', 'duration'])

# 2. Пустые страны заменим на Unknown
df['main_production'] = df['main_production'].fillna('Unknown')
df['main_genre'] = df['main_genre'].fillna('Unknown')

# 3. В числовых полях заменим неподходящие типы
df['score'] = pd.to_numeric(df['score'], errors='coerce')
df['release_year'] = pd.to_numeric(df['release_year'], errors='coerce')
df['number_of_votes'] = pd.to_numeric(df['number_of_votes'], errors='coerce')
df['duration'] = pd.to_numeric(df['duration'], errors='coerce')

# 4. Пропуски в числовых полях заполняем 0
df['score'] = df['score'].fillna(0)
df['number_of_votes'] = df['number_of_votes'].fillna(0)

# Проверка после обработки
print("\nПропуски после обработки:")
print(df.isna().sum())

Количество пропусков в каждом столбце:
title              5
release_year       2
score              3
number_of_votes    1
duration           0
main_genre         0
main_production    1
dtype: int64

Пропуски после обработки:
title              0
release_year       0
score              0
number_of_votes    0
duration           0
main_genre         0
main_production    0
dtype: int64


Проверка на пропуски выявила некорректные значения. Все найденные несоответствия и ошибки были успешно обработаны путём удаления некорректных строк при помощи метода 'dropna()' и замены пустых значений методом 'fillna()' с предварительной заменой некоторых типов при помощи метода 'to_numeric()'.

Строки с пустым названием фильма, годом выпуска или длительностью фильма удалены, пустые страны заменены на Unknown, а в числовых полях заменены неподходящие типы и пустые значения заменены на 0.

#8. Проверка на явные и неявные дубликаты

In [68]:
# Проверка на явные дубликаты
print(f"Количество явных дубликатов: {df.duplicated().sum()}")

# Удаление явных дубликатов
df = df.drop_duplicates().reset_index(drop=True)
print(f"Количество дубликатов после удаления: {df.duplicated().sum()}")

# Проверка на неявные дубликаты в столбцах
print("\nПроверка неявных дубликатов в столбце 'main_genre':")
print(df['main_genre'].value_counts())

# Приведение жанров к нижнему регистру
df['main_genre'] = df['main_genre'].str.lower()

# Проверка уникальных значений после приведения
print("\nУникальные жанры после нормализации:")
print(df['main_genre'].unique())

print("\nПроверка неявных дубликатов в столбце 'main_production':")
print(df['main_production'].value_counts())

# Приведение стран к верхнему регистру
df['main_production'] = df['main_production'].str.upper()

print("\nУникальные страны после нормализации:")
print(df['main_production'].unique())

Количество явных дубликатов: 2
Количество дубликатов после удаления: 0

Проверка неявных дубликатов в столбце 'main_genre':
main_genre
drama          147
thriller        59
comedy          55
crime           21
documentary     20
romance         20
fantasy         19
horror           8
western          7
scifi            7
action           5
war              3
animation        3
musical          3
COMEDY           2
sports           1
dramaa           1
Name: count, dtype: int64

Уникальные жанры после нормализации:
['documentary' 'scifi' 'drama' 'comedy' 'western' 'action' 'crime'
 'thriller' 'war' 'fantasy' 'romance' 'horror' 'animation' 'sports'
 'dramaa' 'musical']

Проверка неявных дубликатов в столбце 'main_production':
main_production
US    151
IN    115
GB     29
JP      9
DE      9
FR      7
TR      7
ES      6
CA      4
KR      4
IT      4
CN      3
AU      3
BE      2
NO      2
NZ      2
HU      2
UA      2
HK      2
NL      2
IE      2
CD      1
ZA      1
MX      1
AR      

Проверка выявила 2 явных дубликата, дубли были удалены при помощи метода 'drop_duplicates()' и проиндексированны при помощи метода 'reset_index(drop=True)'.

Также были выявлены неявные дубликаты в жанре. Дубликаты были устранены путём приведения данных в колонке к нижнему регистру.

#9. Финальная проверка датасета

In [69]:
# Проверка текущих типов данных
print("Общая информация:")
df.info()
print("\nТекущие типы данных:")
print(df.dtypes)
print("\nТекущие пропуски:")
print(df.isna().sum())

Общая информация:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 381 entries, 0 to 380
Data columns (total 7 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   title            381 non-null    object 
 1   release_year     381 non-null    float64
 2   score            381 non-null    float64
 3   number_of_votes  381 non-null    float64
 4   duration         381 non-null    int64  
 5   main_genre       381 non-null    object 
 6   main_production  381 non-null    object 
dtypes: float64(3), int64(1), object(3)
memory usage: 21.0+ KB

Текущие типы данных:
title               object
release_year       float64
score              float64
number_of_votes    float64
duration             int64
main_genre          object
main_production     object
dtype: object

Текущие пропуски:
title              0
release_year       0
score              0
number_of_votes    0
duration           0
main_genre         0
main_production    0
dtype: int

Ранее на шагах 7 и 8 все выявленные несоответствия типов, пустые значения, пропуски и дубли были устранены.

При итоговой проверке информации в датасете и проверке типов данных никаких несоответствий уже не выявлено.


#10. Выводы по работе

## Выводы по анализу данных

В ходе работы был проведен предварительный анализ данных датасета.

Исходный датасет содержит информацию о фильмах с 7 столбцами.

При анализе датасета были выявлены и устранены проблемы с пропусками и дубликатами данных, также данные датасета были приведены к соответствующим типам данных.

Для анализа и обработки данных использовалась библиотека '**pandas**'.

Данная библиотека предназначена для работы с табличными данными и имеет множество инструментов для анализа.

В ходе работы были применены такие возможности библиотеки как:

- Чтение и представление табличных данных
- Переименовывание столбцов таблицы
- Поиск и уничтожение дубликатов данных
- Работа с типами данных
- Поиск и заполнение пропущенных значений в таблице.

Благодаря применению возможностей библиотеки были найдены и решены некоторые проблемы с данными в исходной таблице.

В таблице ниже представлены найденные проблемы и их решение.

| Проблема | Решение |
|----------|---------|
| Пропущенные значения в столбцах | Заполнены 0 и 'Unknown'|
| Пустые названия фильмов | Строки удалены |
| Явные дубликаты строк | Удалены методом drop_duplicates() |
| Неявные дубликаты в жанрах | Приведены к нижнему регистру |
| Несоответствие типов данных | Преобразованы в корректные типы|

Применение возможностей библиотеки позволило добиться следующих результатов в плане предподготовки данных:

- Все числовые столбцы имеют корректные типы данных
- Данные в столбцах приведены к единому формату
- Пропуски устранены или заполнены
- Дубликаты удалены

В результате проделанной работы исходный датасет стал пригодным для дальнейшей обработки и анализа.